File: web/src/pages/Home/components/Log.ts
import z from 'zod'

export const logEntrySchema = z.object({
    time: z.coerce.date(),
    level: z.enum(['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']),
    request_id: z.string(),
    title: z.string(),
    service_name: z.string().optional(),
    controller_name: z.string().optional(),
    message: z.record(z.string()).optional(),
    additional_data: z.record(z.string()).optional(),
})


export type LogEntry = z.infer<typeof logEntrySchema>

export default class Log {
	constructor(
		readonly time: Date,
		readonly level: "DEBUG" | "INFO" | "WARNING" | "ERROR" | "CRITICAL",
		readonly request_id: string,
		readonly title: string,
		readonly service_name?: string,
		readonly controller_name?: string,
		readonly message?: Record<string, string>,
		readonly additional_data?: Record<string, string>
	) {}

	static fromJSON(json: LogEntry) {
		
		return new Log(
			json.time, 
			json.level, 
			json.request_id, 
			json.title, 
			json.service_name, 
			json.controller_name, 
			json.message, 
			json.additional_data
		)
	}
}

File: server/src/data_layer/model/log.rs
use chrono::{DateTime, Utc};
use diesel::{deserialize::Queryable, prelude::Insertable, query_builder::AsChangeset, Selectable};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::prelude::FromRow;
use utoipa::ToSchema;
use uuid::Uuid;
use derive_more::Display;

#[derive(sqlx::Type, Display, Debug, Serialize, Deserialize, diesel_derive_enum::DbEnum, ToSchema, Clone, Copy)]
#[ExistingTypePath = "crate::schema::sql_types::Loglevel"]
#[serde(rename_all = "lowercase")]
pub enum Loglevel {
    DEBUG,
    INFO,
    WARNING,
    ERROR,
    CRITICAL,
}


#[derive(FromRow, Debug, Serialize, Deserialize, Queryable, Selectable, Insertable, AsChangeset, ToSchema)]
#[diesel(table_name = crate::schema::logs)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Log {
    #[schema(value_type = String, format = "date-time")]
    pub time: DateTime<Utc>,
    pub level: Loglevel,
    pub request_id: Uuid,
    pub title: String,
    pub service_name: Option<String>,
    pub controller_name: Option<String>,
    pub message: Option<Value>,
    pub additional_data: Option<Value>,
}
